include/ts/DbgCtl.h

```diff
@@ -0,0 +1,120 @@
+/** @file
+
+  DbgCtl class header file.
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#pragma once
+
+#include <atomic>
+#include <utility>
+
+#include <ts/apidefs.h> // For TS_PRINTFLIKE
+
+class DiagsConfigState;
+
+class DbgCtl
+{
+public:
+  // Tag is a debug tag.  Debug output associated with this control will be output when debug output
+  // is enabled globally, and the tag matches the configured debug tag regular expression.
+  //
+  DbgCtl(char const *tag) : _ptr{_new_reference(tag)} {}
+
+  ~DbgCtl() { _rm_reference(); }
+
+  bool
+  tag_on() const
+  {
+    return _ptr->second;
+  }
+
+  char const *
+  tag() const
+  {
+    return _ptr->first;
+  }
+
+  bool
+  on() const
+  {
+    auto m{_config_mode.load(std::memory_order_relaxed)};
+    if (!m) {
+      return false;
+    }
+    if (!_ptr->second) {
```

Comment:
Technically `second` should be std::atomic (or at least volatile), since it can be written from another thread.  But doing that risks causing the generation of less performant code.  And, it seems unlikely that any register buffering of the value would cause actual problems.